File: November2022/Java/InsertDeleteGetRandomO1.java
// @saorav21994

class RandomizedSet {
    int count = 0;
    HashMap<Integer, Integer> map;
    public RandomizedSet() {
        map = new HashMap<>();
    }
    
    public boolean insert(int val) {
        if (!map.containsKey(val)) {
            map.put(val, 1);
            return true;
        }
        return false;
    }
    
    public boolean remove(int val) {
        if (map.containsKey(val)) {
            map.remove(val);
            return true;
        }
        return false;
    }
    
    public int getRandom() {
        Object[] keys = map.keySet().toArray();
        Random rand = new Random();
        return (int)keys[rand.nextInt(keys.length)];
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: November2022/C++/insert-delete-getrandom-o1.cpp
class RandomizedSet {
private:
    vector<int>num;
    map<int,int>index;
public:
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if(index.find(val)!=index.end()){
         
            return false;
        }
        
           index[val] = num.size();
            num.push_back(val);
        
        return true;
    }
    
    bool remove(int val) {
         if(index.find(val)==index.end()){
         
            return false;
        }
        
          int ind =  index[val];
           num[ind] = num[num.size()-1];
        index[num[num.size()-1]] = ind;
        num.pop_back();
        index.erase(val);
        return true;
    }
    
    int getRandom() {
        int ind = rand()%num.size();
        
        return num[ind];
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */